test/bypass_test.exs

```diff
@@ -449,45 +451,44 @@ defmodule BypassTest do
   closed error and not a failed to connect error, when we test Bypass.down
   """
   def request(port, path \\ "/example_path", method \\ :post) do
-    {:ok, conn} = :gun.start_link(self(), '127.0.0.1', port, %{retry: 0})
-
-    try do
-      case :gun.await_up(conn, 250) do
-        {:ok, _protocol} ->
-          stream =
-            case method do
-              :post -> :gun.post(conn, path, [], "")
-              :get -> :gun.get(conn, path, [])
-            end
-
-          case :gun.await(conn, stream, 250) do
-            {:response, :fin, status, _headers} ->
-              {:ok, status, ""}
-
-            {:response, :nofin, status, _headers} ->
-              case :gun.await_body(conn, stream, 250) do
-                {:ok, data} -> {:ok, status, data}
-                {:error, _} = error -> error
-              end
-
-            {:error, _} = error ->
-              error
-          end
+    method = method |> Atom.to_string() |> String.upcase()
 
-        {:error, :timeout} ->
-          raise "Expected gun to die, but it didn't."
+    with {:ok, conn} <- Mint.HTTP.connect(:http, "127.0.0.1", port),
+         {:ok, conn, ref} = Mint.HTTP.request(conn, method, path, [], "") do
+      result = receive_responses(conn, ref)
+      {:ok, _conn} = Mint.HTTP.close(conn)
+      result
+    end
+  end
 
-        {:error, :normal} ->
-          # `await_up` monitors gun and errors only if gun died (or after `timeout`). That happens
-          # when gun can't connect and is out of retries (immediately in our case) so we know that
-          # gun is dead.
-          {:error, :noconnect}
-      end
-    after
-      Process.unlink(conn)
-      monitor = Process.monitor(conn)
-      Process.exit(conn, :kill)
-      assert_receive {:DOWN, ^monitor, :process, ^conn, _}
+  defp receive_responses(conn, ref, result \\ %{done: false, status: 100, body: []}) do
+    receive do
+      message ->
+        with {:ok, conn, responses} <- Mint.HTTP.stream(conn, message),
+             %{} = result <-
+               Enum.reduce(responses, result, fn response, result ->
+                 case response do
+                   {:status, ^ref, status} ->
+                     %{result | status: status}
+
+                   {:headers, ^ref, _headers} ->
+                     result
+
+                   {:data, ^ref, data} ->
+                     %{result | body: result.body ++ [data]}
```

Comment:
This can become quite expensive (the prepending of data) so it is better to append and then reverse at the end. Here is a version of `receive_responses` that addresses this and is tail recursive:

```elixir
  defp receive_responses(conn, ref, status, body) do
    receive do
      message ->
        with {:ok, conn, responses} <- Mint.HTTP.stream(conn, message) do
          receive_responses(responses, conn, ref, status, body)
        end
    end
  end

  defp receive_responses([], conn, ref, status, body) do
    receive_responeses(conn, ref, status, body)
  end

  defp receive_responses([response | responses], conn, ref, status, body) do
    case response do
      {:status, ^ref, status} ->
        receive_responses(responses, conn, ref, status, body)

      {:headers, ^ref, _headers} ->
        receive_responses(responses, conn, ref, status, body)

      {:data, ^ref, data} ->
        receive_responses(responses, conn, ref, status, [data | body])

      {:done, ^ref} ->
        _ = Mint.HTTP.close(conn)
        {:ok, status, body |> Enum.reverse() |> IO.iodata_to_binary()}

      {:error, ^ref, _reason} = error ->
        error
    end
  end
```

You just need to call it as `receive_responses(conn, ref, 100, [])`.

Reply:
Thanks a lot! 😊 I thought about tail recursion for a second, but then just skipped it because it's only an internal helper function. But as bypass is quite popular in the Elixir community, and people also read code in test cases, I guess writing proper code and leading by example is the best!

Thx again, will change it ASAP! 😊